src/common.ts

```diff
@@ -0,0 +1,107 @@
+import { Address, BigInt, ethereum } from "@graphprotocol/graph-ts"
+import { ERC20 } from "../generated/UniswapV2Factory/ERC20"
+import {
+    Token,
+    RewardToken,
+    DexAmmProtocol,
+    LiquidityPool,
+} from "../generated/schema"
+import {
+    Network,
+    ProtocolType,
+    RewardTokenType,
+} from "./constants"
+
+const getUuid = require('uuid-by-string')
+
+
+export function getOrCreateERC20Token(event: ethereum.Event, address: Address): Token {
+    let addressHex = address.toHexString()
+    let token = Token.load(addressHex)
+    if (token != null) {
+        return token as Token
+    }
+
+    token = new Token(addressHex)
+    let tokenInstance = ERC20.bind(address)
+    let tryName = tokenInstance.try_name()
+    if (!tryName.reverted) {
+        token.name = tryName.value
+    }
+    let trySymbol = tokenInstance.try_symbol()
+    if (!trySymbol.reverted) {
+        token.symbol = trySymbol.value
+    }
+    let tryDecimals = tokenInstance.try_decimals()
+    if (!tryDecimals.reverted) {
+        token.decimals = tryDecimals.value
+    }
+    token.save()
+
+    return token as Token
+}
+
+export function getOrCreateERC20RewardToken(event: ethereum.Event, address: Address): RewardToken {
+    let addressHex = address.toHexString()
+    let rewardToken = RewardToken.load(addressHex)
+    if (rewardToken != null) {
+        return rewardToken as RewardToken
+    }
+
+    rewardToken = new RewardToken(addressHex)
+    let rewardTokenInstance = ERC20.bind(address)
+    let tryName = rewardTokenInstance.try_name()
+    if (!tryName.reverted) {
+        rewardToken.name = tryName.value
+    }
+    let trySymbol = rewardTokenInstance.try_symbol()
+    if (!trySymbol.reverted) {
+        rewardToken.symbol = trySymbol.value
+    }
+    let tryDecimals = rewardTokenInstance.try_decimals()
+    if (!tryDecimals.reverted) {
+        rewardToken.decimals = tryDecimals.value
+    }
+    rewardToken.type = RewardTokenType.DEPOSIT
+    rewardToken.save()
+
+    return rewardToken as RewardToken
+}
+
+export function getOrCreateEthereumDexAmmProtocol(name: string, slug: string): DexAmmProtocol {
+    let uuidHash = getUuid(slug);  // Hash the slug into an UUID
```

Comment:
I assume that the "slug" is globally unique, thus I'm using its hash as the ID for the protocol here. Is this a fair assumption?

If so -- should we simply use "slug" as the primary key instead of "ID"?

Another question is what should the name and slug be for Sushi. Given the the core protocol is a fork of Uniswap V2, would it be more approriate to set the name/slug to "Uniswap V2"/"uniswap-v2" or "Sushiswap"/"sushiswap"?

Reply:
Let's use the address of the factory contract for ID.

Reply:
Sushiswap/sushiswap is good!